docs/developer/gvisor.md

```diff
@@ -0,0 +1,293 @@
+which will also have two layers:
+* The **outer** container image will contain just Python3 and `runsc`, the
+  latter downloaded from the official gVisor website. It will also contain an
+  entrypoint that will launch `runsc`. Finally, it will contain the **inner**
+  container image (see below) as filesystem clone under
+  `/dangerzone-image/rootfs`.
+* The **inner** container image is practically the original Dangerzone image, as
+  we've always built it, which contains the necessary tooling to rasterize a
+  document.
+
+### Spawning the container
+
+Spawning the container now becomes a multi-stage process:
+
+The `Container` isolation provider spawns the container as before, with the
+following changes:
+
+* It adds two Linux capabilities to the **outer** container that didn't exist
+  before: `SETFCAP` and `SYS_CHROOT`. Those capabilities are necessary to run
+  `runsc` rootless, and are not inherited by the **inner** container.
+* It removes the `--userns keep-id` argument, which mapped the user outside the
+  container to the same UID (normally `1000`) within the container. This was
+  originally required when we were mounting host directories within the
+  container, but this no longer applies to the gVisor integration. By removing
+  this flag, the host user maps to the root user within the container (UID `0`).
+  - In distributions that offer Podman version 4 or greater, we use the
+    `--userns nomap` flag. This flag greatly minimizes the attack surface,
+    since the host user is not mapped within the container at all.
+* In distributions that offer Podman 3.x, we add a seccomp filter that adds the
+  `ptrace` syscall, which is required for running gVisor.
+
+Then, the following happens when Podman/Docker spawns the container:
+
+1. _(outer container)_ The entrypoint code finds from `sys.argv` the command
+   that Dangerzone passed to the `docker run` / `podman run` invocation.
+   Typically, this command is:
+
+   ```
+   /usr/bin/python3 -m dangerzone.conversion.doc_to_pixels
+   ```
+
+2. _(outer container)_ The entrypoint code then creates an OCI config for
+   `runsc` with the following properties:
+   * Use UID/GID 1000 in the **inner** container image.
+   * Run the command we detected on step 1.
+   * Enable only the following Linux capabilities in the **inner** container
+     image: `CAP_SETUID` and `CAP_SETGID`. These capabilities can be removed
+     once https://github.com/google/gvisor/issues/9918 is fixed, but the
+     important thing is that they won't be passed to the conversion process.
+   * Limit the number of open files to 4096.
+   * Use the `/dangerzone-image/rootfs` directory as the root path for the
+     **inner** container.
+   * Mount a gVisor view of the `procfs` hierarchy under `/proc` , and then
+     mount `tmpfs` in the `/dev`, `/sys` and `/tmp` mount points. This way, no
+     host-specific info may leak to the **inner** container.
```

Comment:
gVisor doesn't mount any host filesystem by default. If the mounts are removed from the OCI config, these directories will simply not exist in the inner container. The mounts in the OCI config for `/proc`, `/dev`, `/sys` etc are mostly just because most Linux applications expect these mounts to exist, rather than being about hiding host-specific info. (Nonetheless, this bullet point is right that there's no leak)

Reply:
I realize now that this rewording was probably due to [my comment here](https://github.com/freedomofpress/dangerzone/commit/8641b66b0db634d1b6b849f9047a93671d7c5a13#r142338345). Sorry if I had emphasized the wrong thing :) I think this bullet point in the .md file is fine as-is, no need to change it.

Reply:
Got it. I was not aware that the directories are not created, but I guess I should have expected it.

Ok, so I'm leaving this bullet point as is. To be honest, I was itching to also write that "conversely, no container data persist to the host", due to the usage of `tmpfs`. This would be the final blow to https://github.com/freedomofpress/dangerzone/issues/633. Still, because `tmpfs` is backed by swap, I don't think we can safely make this assumption.

In the future, we can consider mounting `ramfs` on the LibreOffice dir only (currently `/home/dangerzone`), since it's under our control, and we create a single file there.